core/state/statedb.go

```diff
@@ -773,42 +765,49 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
  */
 
 // AddBalance adds amount to the account associated with addr.
-func (s *StateDB) AddBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) {
+func (s *StateDB) AddBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
 	stateObject := s.getOrNewStateObject(addr)
+	if stateObject == nil {
+		return uint256.Int{}
+	}
 
 	if s.mvHashmap != nil {
 		// ensure a read balance operation is recorded in mvHashmap
 		s.GetBalance(addr)
 	}
 
-	if stateObject != nil {
-		stateObject = s.mvRecordWritten(stateObject)
-		stateObject.AddBalance(amount, reason)
-		MVWrite(s, blockstm.NewSubpathKey(addr, BalancePath))
-	}
+	stateObject = s.mvRecordWritten(stateObject)
+	MVWrite(s, blockstm.NewSubpathKey(addr, BalancePath))
+	return stateObject.AddBalance(amount)
 }
 
 // SubBalance subtracts amount from the account associated with addr.
-func (s *StateDB) SubBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) {
+func (s *StateDB) SubBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
 	stateObject := s.getOrNewStateObject(addr)
+	if stateObject == nil {
+		return uint256.Int{}
+	}
 
 	if s.mvHashmap != nil {
 		// ensure a read balance operation is recorded in mvHashmap
 		s.GetBalance(addr)
 	}
 
-	if stateObject != nil {
-		stateObject = s.mvRecordWritten(stateObject)
-		stateObject.SubBalance(amount, reason)
-		MVWrite(s, blockstm.NewSubpathKey(addr, BalancePath))
+	// TODO(manav): Confirm if we need to record if amount is zero
+	stateObject = s.mvRecordWritten(stateObject)
+	MVWrite(s, blockstm.NewSubpathKey(addr, BalancePath))
+
+	if amount.IsZero() {
```

Comment:
@cffls a new check for zero amount is added. Do we need to capture using `MVWrite` in that case or not? I've kept that call anyways (irrespective of value) as I don't see any harm but just confirming.

Reply:
Yea I think we can keep it as it is.